utilities/general-helpers.metta

```diff
@@ -207,3 +207,153 @@
+;;         it happens like this --> (cons a (cons-atom b (c)))
+;;         the above doesnt work as it is but since we are passing the result of each stage to a variable metta can deal with that
+;;         ! (++ (0 1 2) 3) --> (0 1 2 3)
+;;         ! (++ (0 1 2) ((0 1 3) (1 3))) --> (0 1 2 (0 1 3) (1 3))
+
+(= (++ $expr $new)
+    (if (== $expr ())
+        (if (isLiteral $new) ($new) $new)
+        (let* 
+            (
+                ($first (car-atom $expr))
+                ($tail (cdr-atom $expr))
+                ; (() (println! $first))
+            )
+            (if (== $tail ())
+                (if (isLiteral $new)
+                    (cons-atom $first ($new))
+                    (cons-atom $first $new)
+                )
+                (let $carry (++ $tail $new)
+                    (cons-atom $first $carry)
+                )))))
+
+;; replaceByIndex replaces an atom at a specific index position with $val -- 0 (indexed counting is implied)
+;; How it works
+;;          parameters:- 1. $expr, parent expression 2. $index, the desired index position 3.$val, value to replace the old element 
+;;          as $index position
+;;          1st checks if $expr == (). If so, return the empty expression
+;;          THEN -- take the fist and rest of $expr
+;;          if $expr is a one element expression and the desired index is 0 return ($new)
+;;          if $index not 0 just return the original expression
+;;          if $expression is a multi-element expression
+;;          check if index == 0, if so append the new $val to $acc and append the this to the rest of expression, $tail
+;;          if $index not 0, call the function with (- $index 1) and all the previous elements appended to $acc
+;; Examples 
+;;         ! (replaceByIndex (0 1 2 3 4) 0 7) -->             (7 1 2 3 4)
+;;         ! (replaceByIndex (0 1 2 3 4) 2 7) -->             (0 1 7 3 4)
+;;         ! (replaceByIndex ((0) (0 1) (2 3 4)) 1 (2 1)) --> ((0) (2 1) (2 3 4))
+
+(: replaceByIndex (-> Expression Number $t Expression))
+(= (replaceByIndex $expr $index $new)
+    (if (== $expr ())
+        $expr
+        (let*
+            (
+                ($first (car-atom $expr))
+                ($tail (cdr-atom $expr))
+            )
+            (if (== $index 0)
+                (if (== $tail ())
+                    ($new)
+                    (cons-atom $new $tail)
+                )
+                (let $c (replaceByIndex $tail (- $index 1) $new)
+                    (cons-atom $first $c)
+                )))))
+
+;; Select atom by index
+;; description:-        0 indexed selection of atom in a tuple
+;;                      Returns -1 if index is out of range
+;;                      If $index 0 -- firs element is returned (car-atom $expr)
+;;                      Otherwise calls selectByIndex recusively with the rest of $expr, (cdr-atom $expr) and (- $index 1)
+;; Example
```

Comment:
The example and description isn't necessary for this function, it is pretty much self explanatory. Just add a comment like: "Select atom by index and return -1 if index is out of range".